### PR TITLE
feat: evaluate all oneonone conversations implement [WPB-5812]

### PR DESCRIPTION
**Pull request by markusfassbender**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5812" title="WPB-5812" target="_blank">Sub-taskWPB-5812</a>  [iOS] Implement the logic for evaluating all 1:1 conversations
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

💡 this merges into epic branch #944!

Implements fetching a list of all users and migrating their conversations, if needed.

#### Test Coverage

- [x] I have added automated test to this contribution

#### How to Test

- This will be tested on the epic branch, see ticket.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/Synchronization/Strategies/EvaluateOneOnOneConversationsStrategy.swift [hunk 1]
```diff
@@ -36,15 +38,58 @@ final class EvaluateOneOnOneConversationsStrategy: AbstractRequestStrategy {
...
+                return
+            }
+
+            let syncContext = await self.managedObjectContext.perform { [weak self] in
+                self?.managedObjectContext.zm_sync
+            }
 
-            syncStatus.finishCurrentSyncPhase(phase: syncPhase)
+            guard let syncContext else {
+                await failCurrentSyncPhase(errorMessage: "can not perform strategy without sync context!")
```
Comment (netbe) [inline wire-ios-sync-engine/Source/Synchronization/Strategies/EvaluateOneOnOneConversationsStrategy.swift]:
```suggestion
                await failCurrentSyncPhase(errorMessage: "cannot perform strategy without sync context!")
```

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift [hunk 2]
```diff
@@ -711,17 +711,20 @@ extension ZMUserSession: ZMSyncStateDelegate {
...
         }
 
-        let selfClient = ZMUser.selfUser(in: syncContext).selfClient()
-        if selfClient?.hasRegisteredMLSClient == true {
-            Task {
+        Task {
+            let hasRegisteredMLSClient = await syncContext.perform {
+                let selfClient = ZMUser.selfUser(in: self.syncContext).selfClient()
+                return selfClient?.hasRegisteredMLSClient == true
+            }
```
Comment (netbe) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift]: **question:** why is this moved?

Comment (markusfassbender) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift]: Because we now call `didFinishSlowSync()` also from another thread through async await und thus the access of `ZMUser.selfUser(in: syncContext).selfClient()` crashed.

Comment (netbe) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift]: you should call it from syncContext

Comment (markusfassbender) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift]: Please review my changes about the queues

**Review by David-Henner — APPROVED** [review]

File: wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift [hunk 3]
```diff
@@ -62,12 +65,29 @@ public final class OneOnOneResolver: OneOnOneResolverInterface {
 
     // MARK: - Methods
 
+    public func resolveAllOneOnOneConversations(in context: NSManagedObjectContext) async throws {
+        let users: [ZMUser] = try await context.perform {
+            let request = NSFetchRequest<ZMUser>(entityName: ZMUser.entityName())
+            request.predicate = NSPredicate(format: "oneOnOneConversation != nil")
```
Comment (netbe) [inline wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift]: suggestion: move predicate to ZMUser.
and use keypath:
```
 NSPredicate(
            format: "%K == nil,
            #keyPath(ZMUser.oneOnOneConversation)
```

File: wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift [hunk 4]
```diff
@@ -62,12 +65,29 @@ public final class OneOnOneResolver: OneOnOneResolverInterface {
...
+            return try context.fetch(request)
+        }
+
+        for user in users {
+            guard let userID = await context.perform({ user.qualifiedID }) else {
+                WireLogger.conversation.error("required to have a user's qualifiedID to resolve 1-1 conversation!")
+                throw OneOnOneResolverError.otherUserNotFound
```
Comment (netbe) [inline wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift]: do we stop the full process if we can't find one userID? should we continue instead to other users?

File: wire-ios-sync-engine/Source/Synchronization/Strategies/EvaluateOneOnOneConversationsStrategy.swift [hunk 5]
```diff
@@ -36,15 +40,52 @@ final class EvaluateOneOnOneConversationsStrategy: AbstractRequestStrategy {
...
-            // - Loop through the list and evaluate the 1:1 conversation for each user.
+        guard isSyncing, task == nil else {
+            return nil
+        }
+
+        WireLogger.conversation.info("EvaluateOneOnOneConversationsStrategy: start evaluate one on one conversations!")
+
+        // store task to avoid duplicated entries and concurrency issues
+        task = Task { [weak self] in
+            guard let self, let syncContext = managedObjectContext.performAndWait({ self.managedObjectContext.zm_sync }) else {
```
Comment (netbe) [inline wire-ios-sync-engine/Source/Synchronization/Strategies/EvaluateOneOnOneConversationsStrategy.swift]: what is the value of managedObjectContext? syncContext already, no?

**Review by netbe — COMMENTED** [review]
I still have some open questions to check

File: wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift (hunk 4 above)
Comment (markusfassbender) [reply wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift]: Good catch, I have changed it

File: wire-ios-sync-engine/Source/Synchronization/Strategies/EvaluateOneOnOneConversationsStrategy.swift (hunk 5 above)
Comment (markusfassbender) [reply wire-ios-sync-engine/Source/Synchronization/Strategies/EvaluateOneOnOneConversationsStrategy.swift]: this is an implicit assumption, that was not clear to me.. did some change but added `precondition` to make it better understandable for the reader.

File: wire-ios-data-model/Tests/OneOnOne/OneOnOneResolverTests.swift [hunk 6]
```diff
@@ -20,72 +20,182 @@ import XCTest
 @testable import WireDataModel
 @testable import WireDataModelSupport
 
-final class OneOnOneResolverTests: ZMBaseManagedObjectTest {
+final class OneOnOneResolverTests: XCTestCase {
```
Comment (agisilaos) [inline wire-ios-data-model/Tests/OneOnOne/OneOnOneResolverTests.swift:23]: Thank you! 👍

**Review by agisilaos — COMMENTED** [review]
Left a suggestion!

File: wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift [hunk 7]
```diff
@@ -203,37 +205,53 @@ extension ConnectionRequestStrategy: ZMEventConsumer {
...
-                                WireLogger.conversation.error("Error resolving one-on-one conversation: \(error)")
-                                assertionFailure("Error resolving one-on-one conversation: \(error)")
-                            }
-                        }
-                    }
-
+                guard let payload = Payload.UserConnectionEvent(payloadData) else {
+                    return
                 }
+                processEvent(payload)
```
Comment (agisilaos) [inline wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift]: processEvent seems a bit too general, how about processUserConnectionEvent?

```suggestion
    processUserConnectionEvent(payload)
```

File: wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift [hunk 8]
```diff
@@ -203,37 +205,53 @@ extension ConnectionRequestStrategy: ZMEventConsumer {
...
 
             default:
                 break
             }
         }
     }
 
+    private func processEvent(_ payload: Payload.UserConnectionEvent) {
```
Comment (agisilaos) [inline wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift]: processEvent seems a bit too general, how about processUserConnectionEvent?

```suggestion
    private func processUserConnectionEvent(_ payload: Payload.UserConnectionEvent) {
```

Comment (markusfassbender) [reply wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift]: thanks done
